### DISCUSSION TITLE
How to run action itself in that action?

### DISCUSSION THREAD

**Discussion by baeharam** [mobxjs/mobx-state-tree#1943] [Q&A] [answered]
THE SITUATION

1) I have async function to fetch something, let's called A function.
2) When this function gets authentication error, it shoulde be authenticated by auth API called B function.
3) After B is called, A doesn't get authentication error because user is authenticated by B.
4) Thus, I have to call A again to get proper response.

To implement this process, I have to call action A in action A (`flow`).

```typescript
types.model("Example")
.actions((self) => ({
  fetcherA: flow(function* (){
    // ...some logic
   try {
    // ...some logic
   } catch (error) {
     if (error === AUTH_ERROR) {
       fetcherB().then(_ => fetcherA()) // I have to call retry action(fetcherA) itself
     }
   }
  })
}));
```

How can I do this? Thanks.

**Comment by EmilTholin** [comment, marked as answer]
Hi @baeharam!

You could put your flows in standalone variables in the scope first and return them afterwards. This way you can references them.

**Example**
```js
types.model("Example").actions((self) => {
  const fetcherB = flow(function* () {
    // ...
  });
  const fetcherA = flow(function* () {
    // ...
    try {
      // ...
    } catch (error) {
      if (error === AUTH_ERROR) {
        yield fetcherB();
        yield fetcherA();
      }
    }
  });

  return {
    fetcherA,
    fetcherB,
  };
});
```

**Reply by baeharam** [reply]
@EmilTholin Thanks for your quick response!
